docs/automated_testing.md

```diff
@@ -0,0 +1,509 @@
+
+### Starting Step
+
+Leave out the extension of the file. Use just the name by itself.
+```
+    Given I start the interview at "yaml_file_name"
+```
+<!-- Given I start the interview at "filename" in lang "Español" -->
+<!-- And I am using a mobile -->
+
+### Observe things about the page
+
+To check the id, look at the YAML `question` block and copy the id from there. Adding this Step into the code of the test can help you, as a human, keep track of what fields you should be filling in next. It will also show up in the logs of the tests and can help you see where things went wrong.
+```
+    Then the question id should be "some yaml block id!"
+```
+```
+    Then I can't continue
+```
+```
+    Then I will be told an answer is invalid
+```
+```
+    Then I arrive at the next page
+```
+
+Screenshots will be in the GitHub action's [artifacts](https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/downloading-workflow-artifacts).
+<!-- And I take a screenshot ?(?:named "([^"]+)")? -->
+```
+    Then I take a screenshot
+```
+```
+    Then I should see the link to "a-url.com"
+```
+
+<!-- Then an element should have the id "some_HTML_element_id" -->
+
+Checking phrases will be language specific. Also, docassemble sometimes uses some weird characters on screen, so copying it straight from the screen is best.
+```
+    Then I SHOULD see the phrase "some phrase"
+```
+```
+    Then I should NOT see the phrase "some phrase"
+```
+
+<!-- Then the "a" link leads to "a" -->
+<!-- Then the "a" link opens a working page -->
+<!-- Then the "a" link opens in a new window -->
+<!-- Then the "a" link opens in the same window -->
+
+
+### Set fields
+
+```
+    When I tap to continue
+```
+
+Use this Step to set the values of fields. Comparing this to the story table described above, the first quotes contain the equivalent of the `var` column and the second quotes contain the `value` you want to set.
+```
+    When I set the variable "users[i].hair_color" to "blue"
+```
+
+Sign on a signature field. All signatures are the same.
+```
+    When I sign
+```
+
+Specifically for the Document Assembly Line 4-part name questions. No punctuation. We recommend you just use 2 names - the first name and last name - but you can have all these formats:
+
+- Firstname Lastname
+- Firstname Middlename Lastname
+- Firstname Middlename Lastname Suffix (where suffix is one of the dropdown suffix choices, like `II`)
+
+```
+    When I set the name of "x[i]" to "Ulli User"
+```
+
+Specifically for the Document Assembly Line 5-field address questions. t can be any address you want that matches the format shown, commas and all.
+
+```
+    When I set the address of "users[0]" to "120 Tremont Street, Unit 1, Boston, MA 02108"
+```
+
+The Step for the story table, which is better described in sections above.
+```
+    I get to the question id "some yml block id" with this data:
+```
+
+### Other actions
+
+```
+    When I tap to continue
+```
```

Comment:
This one's duplicated?

Reply:
Yes. I wasn't sure whether to duplicate it or not - duplication is generally bad, but it's one I've used a lot under some circumstances, so I want it to be easy to find and it does seem to be to belong in both categories. What do you think?

Reply:
That makes sense. IMO the design of this section on Steps might need some more work, but you're right, it's probably fine for now.